diff --git a/rakudo/Configure.pl b/rakudo/Configure.pl
index 5b2f034..7cab8fb 100644
--- a/rakudo/Configure.pl
+++ b/rakudo/Configure.pl
@@ -224,6 +224,7 @@ MAIN: {
 
             $config{'nqp_prefix'}    = $nqp_config{'jvm::prefix'};
             $config{'nqp_jars'}      = $nqp_config{'jvm::runtime.jars'};
+            $config{'bld_nqp_jars'}  = join( $config{'cpsep'}, map { $config{'sysroot'} . $_ } split( $config{'cpsep'}, $nqp_config{'jvm::runtime.jars'} ) );
             $config{'nqp_classpath'} = $nqp_config{'jvm::runtime.classpath'};
             $config{'j_runner'}      = $win ? 'perl6-j.bat' : 'perl6-j';
 
diff --git a/rakudo/tools/build/Makefile-JVM.in b/rakudo/tools/build/Makefile-JVM.in
index 617a51a..868e9bf 100644
--- a/rakudo/tools/build/Makefile-JVM.in
+++ b/rakudo/tools/build/Makefile-JVM.in
@@ -11,8 +11,10 @@ J_BUILD_DIR = gen/jvm
 PERL6_LANG_DIR = $(PREFIX)/share/perl6
 
 NQP_JARS = @nqp_jars@
+BLD_NQP_JARS = @bld_nqp_jars@
 
-J_RUN_PERL6 = $(JAVA) -Xss1m -Xms500m -Xmx2000m -Xbootclasspath/a:.@cpsep@$(NQP_JARS)@cpsep@rakudo-runtime.jar@cpsep@perl6.jar -cp @nqp_classpath@ perl6
+J_RUN_NQP_RR = $(JAVA) -Xss1m -Xms500m -Xmx2000m -Xbootclasspath/a:.@cpsep@$(BLD_NQP_JARS)@cpsep@rakudo-runtime.jar -cp $(SYSROOT)@nqp_classpath@ nqp
+J_RUN_PERL6 = $(JAVA) -Xss1m -Xms500m -Xmx2000m -Xbootclasspath/a:.@cpsep@$(BLD_NQP_JARS)@cpsep@rakudo-runtime.jar@cpsep@perl6.jar -cp $(SYSROOT)@nqp_classpath@ perl6
 
 RUNTIME_JAVAS = src/vm/jvm/runtime/org/perl6/rakudo/*.java
 
@@ -248,7 +250,7 @@ j-all: $(PERL6_JAR) $(SETTING_JAR) $(J_RUNNER) lib/Test.pm.jar lib/lib.pm6.jar l
 
 $(RUNTIME_JAR): $(RUNTIME_JAVAS)
 	$(PERL) -MExtUtils::Command -e mkpath bin
-	$(JAVAC) -source 1.7 -cp $(NQP_JARS) -g -d bin -encoding UTF8 $(RUNTIME_JAVAS)
+	$(JAVAC) -source 1.7 -cp $(BLD_NQP_JARS) -g -d bin -encoding UTF8 $(RUNTIME_JAVAS)
 	$(JAR) cf0 rakudo-runtime.jar -C bin/ .
 
 $(PERL6_ML_JAR): src/Perl6/ModuleLoader.nqp src/vm/jvm/ModuleLoaderVMConfig.nqp src/vm/jvm/Perl6/JavaModuleLoader.nqp
@@ -299,8 +301,7 @@ $(PERL6_M_JAR): $(J_METAMODEL_SOURCES) $(PERL6_OPS_JAR)
 
 $(PERL6_B_JAR): $(BOOTSTRAP_SOURCES) $(PERL6_M_JAR)
 	$(J_NQP) $(J_GEN_CAT) $(BOOTSTRAP_SOURCES) > $(J_BUILD_DIR)/BOOTSTRAP.nqp
-	$(PERL) tools/build/nqp-jvm-rr.pl $(J_NQP) --target=jar \
-        --output=$(PERL6_B_JAR) --encoding=utf8 $(J_BUILD_DIR)/BOOTSTRAP.nqp
+	$(J_RUN_NQP_RR) --target=jar --output=$(PERL6_B_JAR) --encoding=utf8 $(J_BUILD_DIR)/BOOTSTRAP.nqp
 
 $(SETTING_JAR): $(PERL6_JAR) $(PERL6_B_JAR) $(J_CORE_SOURCES)
 	$(J_NQP) $(J_GEN_CAT) $(J_CORE_SOURCES) > $(J_BUILD_DIR)/CORE.setting
@@ -344,12 +345,15 @@ lib/Pod/To/Text.pm.jar: lib/Pod/To/Text.pm $(PERL6_JAR) $(SETTING_JAR) $(J_RUNNE
 lib/Pod/To/Text.pm.jar: lib/Pod/To/Text.pm $(PERL6_JAR) $(SETTING_JAR) $(J_RUNNER)
 	.@slash@$(J_RUNNER) --target=jar --output=lib/Pod/To/Text.pm.jar lib/Pod/To/Text.pm
 
+eval-client.pl:
+	$(CP) $(SDKROOT)$(NQP_PREFIX)/bin/eval-client.pl .
+
 ##  testing targets
 j-test    : j-coretest
 
 j-fulltest: j-coretest j-stresstest
 
-j-coretest: j-all
+j-coretest: j-all eval-client.pl
 	$(J_HARNESS) t/01-sanity
 	$(J_HARNESS) t/03-jvm
 	$(J_HARNESS) tivecall
diff --git a/rakudo/tools/build/Makefile-common.in b/rakudo/tools/build/Makefile-common.in
index bb0a52e..cab8c8a 100644
--- a/rakudo/tools/build/Makefile-common.in
+++ b/rakudo/tools/build/Makefile-common.in
@@ -9,6 +9,7 @@ RM_F    = $(PERL) -MExtUtils::Command -e rm_f
 RM_RF   = $(PERL) -MExtUtils::Command -e rm_rf
 
 SYSROOT= @sysroot@
+SDKROOT= @sdkroot@
 PREFIX = @prefix@
 PERL6_LANG_DIR = $(PREFIX)/share/perl6
 
diff --git a/rakudo/tools/build/create-jvm-runner.pl b/rakudo/tools/build/create-jvm-runner.pl
index aa592e3..9977838 100644
--- a/rakudo/tools/build/create-jvm-runner.pl
+++ b/rakudo/tools/build/create-jvm-runner.pl
@@ -23,15 +23,26 @@ die "Invalid target type $type" unless $type eq 'dev' || $type eq 'install';
 my $cpsep = $^O eq 'MSWin32' ? ';' : ':';
 my $bat   = $^O eq 'MSWin32' ? '.bat' : '';
 
-my $preamble = $^O eq 'MSWin32' ? '@' : '#!/bin/sh
-exec ';
-my $postamble = $^O eq 'MSWin32' ? ' %*' : ' "$@"';
-
+my $nqpdir = File::Spec->catfile($nqpprefix, 'share', 'nqp');
+my $nqplibdir = $^O eq 'MSWin32' ? File::Spec->catfile($nqpdir, 'lib') : File::Spec->catfile('${NQP_DIR}', 'lib');
+my $nqpjars = $^O eq 'MSWin32' ? $thirdpartyjars : join( $cpsep, map { $_ =~ s,$nqpdir,\${NQP_DIR},g; $_ } split($cpsep, $thirdpartyjars) );
 my $bindir = $type eq 'install' ? File::Spec->catfile($prefix, 'bin') : $prefix;
-my $jardir = $type eq 'install' ? File::Spec->catfile($prefix, 'share', 'perl6', 'runtime') : $prefix;
-my $libdir = $type eq 'install' ? File::Spec->catfile($prefix, 'share', 'perl6', 'lib') : 'blib';
-my $nqplibdir = File::Spec->catfile($nqpprefix, 'share', 'nqp', 'lib');
+my $perl6dir = $type eq 'install' ? File::Spec->catfile($prefix, 'share', 'perl6') : $prefix;
+my $jardir = $type eq 'install' ? File::Spec->catfile($^O eq 'MSWin32' ? $perl6dir : '${PERL6_DIR}', 'runtime') : $prefix;
+my $libdir = $type eq 'install' ? File::Spec->catfile($^O eq 'MSWin32' ? $perl6dir : '${PERL6_DIR}', 'lib') : 'blib';
 my $sharedir = File::Spec->catfile($prefix, 'share', 'perl6', 'site', 'lib');
+my $perl6jars = join( $cpsep,
+    $^O eq 'MSWin32' ? $nqpjars : '${NQP_JARS}',
+    File::Spec->catfile($jardir, 'rakudo-runtime.jar'),
+    File::Spec->catfile($jardir, $debugger ? 'perl6-debug.jar' : 'perl6.jar'));
+
+my $preamble = $^O eq 'MSWin32' ? '@' : "#!/bin/sh
+: \${NQP_DIR:=\"$nqpdir\"}
+: \${NQP_JARS:=\"$nqpjars\"}
+: \${PERL6_DIR:=\"$perl6dir\"}
+: \${PERL6_JARS:=\"$perl6jars\"}
+exec ";
+my $postamble = $^O eq 'MSWin32' ? ' %*' : ' "$@"';
 
 sub install {
     my ($name, $command) = @_;
@@ -48,13 +59,8 @@ sub install {
     chmod 0755, $install_to if $^O ne 'MSWin32';
 }
 
-my $bootclasspath = join($cpsep,
-    ($thirdpartyjars,
-    File::Spec->catfile($jardir, 'rakudo-runtime.jar'),
-    File::Spec->catfile($jardir, $debugger ? 'perl6-debug.jar' : 'perl6.jar')));
-    
 my $classpath = join($cpsep, ($jardir, $libdir, $nqplibdir));
-my $jopts = '-noverify -Xms100m -Xbootclasspath/a:' . $bootclasspath 
+my $jopts = '-noverify -Xms100m -Xbootclasspath/a:' . $perl6jars 
           . ' -cp ' . ($^O eq 'MSWin32' ? '"%CLASSPATH%";' : '$CLASSPATH:') . $classpath
           . ' -Dperl6.prefix=' . $prefix
           . ' -Djna.library.path=' . $sharedir
@@ -67,6 +73,4 @@ else {
     install "perl6-j", "java $jopts perl6";
     install "perl6-jdb-server", "java -Xdebug -Xrunjdwp:transport=dt_socket,address=8000,server=y,suspend=y $jopts perl6";
     install "perl6-eval-server", "java -Xmx3000m -XX:MaxPermSize=200m $jopts org.perl6.nqp.tools.EvalServer";
-    cp(File::Spec->catfile($nqpprefix,'bin','eval-client.pl'), '.')
-        or die "Couldn't copy 'eval-client.pl' from $nqpprefix: $!";
 }
